snowfakery/standard_plugins/datasets.py

```diff
@@ -0,0 +1,220 @@
+
+
+def _open_db(db_url):
+    "Internal function for opening the database up."
+    engine = create_engine(db_url)
+    metadata = MetaData(bind=engine)
+    metadata.reflect(views=True)
+    return engine, metadata
+
+
+def sql_dataset(db_url: str, tablename: str = None, mode="linear"):
+    "Open the right SQL Dataset iterator based on the params"
+    assert db_url
+    engine, metadata = _open_db(db_url)
+    tables = {
+        name: value
+        for name, value in metadata.tables.items()
+        if not name.startswith("sqlite")
+    }
+    if tablename:
+        table = tables.get(tablename)
+        if table is None:
+            raise AttributeError(f"Cannot find table: {tablename}")
+    elif len(tables) == 0:
+        raise Exception("Database does not exist or has no tables in it")
+    elif len(tables) == 1:
+        table = next(iter(tables.values()))
+    elif len(tables) > 1:
+        raise Exception(
+            f"Database has multiple tables in it and none was selected: {metadata.tables.keys()}"
+        )
+    if mode == "linear":
+        return SQLDatasetLinearIterator(engine, table)
+    elif mode == "shuffle":
+        return SQLDatasetRandomPermutationIterator(engine, table)
+    else:
+        raise NotImplementedError()
+
+
+class DatasetIteratorBase:
+    """Base class for Dataset Iterators
+
+    Subclasses should implement 'self.restart' which puts an iterator into 'self.results'
+    """
+
+    def __next__(self):
+        try:
+            return next(self.results)
+        except StopIteration:
+            self.restart()
+
+    def start(self):
+        "Initialize the iterator in self.results."
+        raise NotImplementedError()
+
+    def restart(self):
+        "Restart the iterator by assigning to self.results"
+        self.start()
+
+    def close(self):
+        "Subclasses should implement this if they need to clean up resources"
+        pass
+
+
+class SQLDatasetIterator(DatasetIteratorBase):
+    def __init__(self, engine, table):
+        self.connection = engine.connect()
+        self.table = table
+        self.start()
+
+    def start(self):
+        self.results = (
+            PluginResult(dict(row)) for row in self.connection.execute(self.query())
+        )
+
+    def close(self):
+        self.connection.close()
+
+    def query(self):
+        "Return a SQL Alchemy SELECT statement"
+        raise NotImplementedError
+
+
+class SQLDatasetLinearIterator(SQLDatasetIterator):
+    "Iterator that reads a SQL table from top to bottom"
+
+    def query(self):
+        return select([self.table])
+
+
+class SQLDatasetRandomPermutationIterator(SQLDatasetIterator):
+    "Iterator that reads a SQL table in random order"
+
+    def query(self):
+        return select([self.table]).order_by(func.random())
+
+
+class CSVDatasetLinearIterator(DatasetIteratorBase):
+    def __init__(self, datasource: Path):
+        self.datasource = datasource
+        self.file = open(self.datasource)
```

Comment:
Shouldn't this have an implementation of `close` to close the file?

Reply:
Also, this should specify `newline=''` so that the `csv` module can handle newlines inside of quoted csv fields, and it should specify an encoding if you don't want to get cp1252 on Windows.